Fix part of #5070: Display empty answer message in ratio input interaction

theMr17 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes part of #5070

Enables the `submit_answer_button` when the pending answer is empty. Instead of disabling the button, an error message, stating "_**Enter a ratio to continue.**_", is now displayed when the user attempts to submit a blank answer.

`RatioInputInteractionViewTestActivity` is added to accessibility-label exemption as this is a test activity.

https://github.com/oppia/oppia-android/assets/84731134/33c797c5-af44-4eaa-980e-fafaccf72d26

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

theMr17 committed:
Adds empty input error message

theMr17 committed:
Adds check for blank ratio input

theMr17 committed:
Enables submit button when answer is empty

theMr17 committed:
Fixes lint issues

adhiamboperes commented:
@theMr17 is this ready for a full pass or are you looking for a preliminary review on approach? Please convert to ready and assign to me if the former.

theMr17 commented:
@adhiamboperes, please do a preliminary review of the approach. I will proceed to write tests after that.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/RatioExpressionInputInteractionViewModel.kt:
```diff
@@ -46,12 +46,17 @@ class RatioExpressionInputInteractionViewModel private constructor(
           errorOrAvailabilityCheckReceiver.onPendingAnswerErrorOrAvailabilityCheck(
             pendingAnswerError,
-            answerText.isNotEmpty()
+            true
```
Maybe add a comment explaining why this is `true`.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/RatioExpressionInputInteractionViewModel.kt:
```diff
@@ -46,12 +46,17 @@ class RatioExpressionInputInteractionViewModel private constructor(
     errorMessage.addOnPropertyChangedCallback(callback)
     isAnswerAvailable.addOnPropertyChangedCallback(callback)
+
+    errorOrAvailabilityCheckReceiver.onPendingAnswerErrorOrAvailabilityCheck(
+      null,
+      true
```
can you add here the argument names so it is clear what is null/true, so that your code is self-documenting.

adhiamboperes commented on app/src/main/res/values/strings.xml:
```diff
@@ -174,6 +174,7 @@
   <string name="ratio_error_invalid_colons" description="The error message for ratio when there is colons next to each other.">Your answer has two colons (:) next to each other.</string>
   <string name="ratio_error_invalid_size" description="The error message for ratio when there is less number of terms than the creator has specified in customization arg.">Number of terms is not equal to the required terms.</string>
   <string name="ratio_error_includes_zero" description="The error message for ratio when there is 0 term in the ratio.">Ratios cannot have 0 as an element.</string>
+  <string name="ratio_error_empty_input">Enter a ratio to continue</string>
```
```suggestion
  <string name="ratio_error_empty_input">Enter a ratio to continue.</string>
```

adhiamboperes reviewed (changes requested):
@theMr17, I have left some suggestions, but the overall approach looks good to me.

theMr17 committed:
Minor fixes

theMr17 committed:
Adds Ratio input test activity

theMr17 committed:
Fixes lint issues

theMr17 committed:
Moves ratio input tests

theMr17 committed:
Minor fixes

theMr17 committed:
Minor fixes

theMr17 committed:
Minor fixes

theMr17 committed:
Minor fixes

theMr17 committed:
Includes ratio test activity in exemptions

theMr17 committed:
Documents ratio test activity

theMr17 committed:
Removes RatioInputInteractionView from InputInteractionViewTestActivity

theMr17 committed:
Reverts unnecessary manifest file changes

theMr17 commented:
@adhiamboperes PTAL

adhiamboperes reviewed (approved):
Thanks @theMr17! This LGTM. 

Thanks for moving the ratio tests to their own suite as well.

adhiamboperes committed:
Merge branch 'develop' into feat/ratio-ui-empty-answer-support

theMr17 committed:
Merge branch 'develop' into feat/ratio-ui-empty-answer-support

theMr17 committed:
Merge branch 'develop' into feat/ratio-ui-empty-answer-support

adhiamboperes commented:
@BenHenning, this PR is blocked on codeowner, PTAL.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/testing/RatioInputInteractionViewTestActivityTest.kt:
```diff
@@ -0,0 +1,471 @@
+    )
+    activityScenario.onActivity { activity ->
+      val pendingAnswer = activity.ratioExpressionInputInteractionViewModel.getPendingAnswer()
+      Truth.assertThat(pendingAnswer.answer).isInstanceOf(InteractionObject::class.java)
```
Please directly import assertThat (for consistency with other tests and per Truth convention).

BenHenning replied:
This has been resolved in the latest changes.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/testing/RatioInputInteractionViewTestActivityTest.kt:
```diff
@@ -0,0 +1,471 @@
+    val activityScenario = ActivityScenario.launch(
+      RatioInputInteractionViewTestActivity::class.java
+    )
+    Espresso.onView(ViewMatchers.withId(R.id.test_ratio_input_interaction_view))
```
Espresso functions like onView & others should also be directly imported. Ditto for ViewMatchers, ViewAssertions, and ViewActions.

BenHenning replied:
This has been resolved in the latest changes.

BenHenning reviewed (changes requested):
Thanks @theMr17! Really sorry about the late review here. I just had two small comments, otherwise the PR LGTM.

theMr17 committed:
Address requested changes

theMr17 committed:
Merge branch 'develop' into feat/ratio-ui-empty-answer-support

theMr17 commented:
@BenHenning PTAL

BenHenning reviewed (approved):
Thanks @theMr17! Please make sure to respond to addressed comments on future PRs. Otherwise the PR LGTM.

theMr17 commented:
> Thanks @theMr17! Please make sure to respond to addressed comments on future PRs. Otherwise the PR LGTM.

Thanks @BenHenning, I will surely keep that in mind.
